File: sdks/typescript/src/bridge/inspection.ts
import { sha256 } from "@noble/hashes/sha2.js";
import { keccak_256 } from "@noble/hashes/sha3.js";

/** Claims are UNAUTHENTICATED. Parsing does not verify the HMAC or transaction binding. */
export interface PreparationProofClaims {
  version: 1;
  keyId: string;
  /** Unix seconds; parsing does not reject expired proofs. */
  expiresAt: number;
  signer: string;
}

function requireValue(ok: unknown, message = "Invalid bridge transaction"): asserts ok {
  if (!ok) throw new Error(message);
}
function unhex(value: string): Uint8Array {
  requireValue(value.length <= 32770 && /^0x(?:[\da-fA-F]{2})+$/.test(value));
  return Uint8Array.from(value.slice(2).match(/../g) ?? [], (b) => Number.parseInt(b, 16));
}
function hex(value: Uint8Array): string {
  return `0x${Array.from(value, (b) => b.toString(16).padStart(2, "0")).join("")}`;
}
function integer(value: Uint8Array): bigint {
  return value.reduce((n, b) => (n << 8n) | BigInt(b), 0n);
}
function word(value: bigint): Uint8Array {
  requireValue(value >= 0n && value <= 0xffffffffffffffffn, "Fuel chain ID must fit u64");
  return unhex(`0x${value.toString(16).padStart(16, "0")}`);
}
function base64url(value: string): Uint8Array {
  requireValue(/^[A-Za-z0-9_-]+$/.test(value), "Invalid preparation proof encoding");
  const bytes = Uint8Array.from(atob(value.replaceAll("-", "+").replaceAll("_", "/")), (c) =>
    c.charCodeAt(0),
  );
  const canonical = btoa(String.fromCharCode(...bytes))
    .replaceAll("+", "-")
    .replaceAll("/", "_")
    .replace(/=+$/, "");
  requireValue(canonical === value, "Noncanonical preparation proof encoding");
  return bytes;
}
/** Decode only. Never use these claims to establish the API's authenticity or signer ownership. */
export function parsePreparationProof(proof: string): PreparationProofClaims {
  requireValue(proof.length <= 2048, "Preparation proof too large");
  const parts = proof.split(".");
  requireValue(
    parts.length === 2 && base64url(parts[1] ?? "").length === 32,
    "Invalid preparation proof",
  );
  const claims = JSON.parse(
    new TextDecoder("utf-8", { fatal: true }).decode(base64url(parts[0] ?? "")),
  ) as PreparationProofClaims;
  requireValue(
    claims &&
      Object.keys(claims).sort().join() === "expiresAt,keyId,signer,version" &&
      claims.version === 1 &&
      typeof claims.keyId === "string" &&
      claims.keyId.length > 0 &&
      Number.isSafeInteger(claims.expiresAt) &&
      claims.expiresAt > 0 &&
      typeof claims.signer === "string" &&
      /^0x([0-9a-f]{40}|[0-9a-f]{64})$/.test(claims.signer),
    "Invalid preparation proof claims",
  );
  return claims;
}

class Reader {
  offset = 0;
  constructor(readonly bytes: Uint8Array) {}
  take(size: number): Uint8Array {
    requireValue(
      Number.isSafeInteger(size) && size >= 0 && size <= this.bytes.length - this.offset,
      "Truncated transaction",
    );
    const result = this.bytes.slice(this.offset, this.offset + size);
    this.offset += size;
    return result;
  }
  num(size = 8): bigint {
    return integer(this.take(size));
  }
  count(max = 16384): number {
    const n = this.num();
    requireValue(n <= BigInt(max), "Unsupported transaction size or field");
    return Number(n);
  }
  padded(size: number): Uint8Array {
    const data = this.take(size);
    requireValue(
      this.take((8 - (size % 8)) % 8).every((b) => b === 0),
      "Nonzero transaction padding",
    );
    return data;
  }
  done(): void {
    requireValue(this.offset === this.bytes.length, "Trailing transaction data");
  }
}

// Bounded canonical RLP, used only for the proxy's unsigned EIP-1559 envelope.
type Rlp = Uint8Array | Rlp[];
function rlp(reader: Reader, depth = 0): Rlp {
  requireValue(depth < 4, "RLP nesting too deep");
  const tag = reader.take(1)[0] ?? 0;
  if (tag < 128) return Uint8Array.of(tag);
  const list = tag >= 192;
  const short = list ? 192 : 128;
  const long = list ? 247 : 183;
  let length = tag - short;
  if (tag > long) {
    const encoded = reader.take(tag - long);
    requireValue(encoded[0] !== 0);
    const n = integer(encoded);
    requireValue(n >= 56n && n <= 16384n);
    length = Number(n);
  }
  const data = reader.take(length);
  if (!list) {
    requireValue(length !== 1 || (data[0] ?? 0) >= 128, "Noncanonical RLP");
    return data;
  }
  const children = new Reader(data);
  const result: Rlp[] = [];
  while (children.offset < data.length) result.push(rlp(children, depth + 1));
  return result;
}
function leaf(value: Rlp | undefined, size?: number): Uint8Array {
  requireValue(
    value instanceof Uint8Array && (size === undefined || value.length === size),
    "Invalid EVM field",
  );
  return value;
}
function evmInt(value: Rlp | undefined): bigint {
  const data = leaf(value);
  requireValue(
    data.length <= 32 && (data.length === 0 || data[0] !== 0),
    "Noncanonical EVM integer",
  );
  return integer(data);
}

export interface EvmDepositInspection {
  type: 2;
  chainId: bigint;
  nonce: bigint;
  messengerAddress: string;
  value: bigint;
  gasLimit: bigint;
  maxFeePerGas: bigint;
  maxPriorityFeePerGas: bigint;
  /** gasLimit * maxFeePerGas, in wei: an upper bound, NOT an actual fee quote.
   * Does not include chain-specific extra fees such as rollup L1 data fees. */
  estimatedNetworkFee: bigint;
  data: string;
  method: "deposit" | "depositWithPermit" | "depositETH";
  recipient: string;
  recipientIsContract: boolean;
  tokenAddress: string | null;
  /** EVM token base units (or wei for depositETH), not API/Fuel base units. */
  amount: bigint;
  permit?: { deadline: bigint; v: number; r: string; s: string };
  /** Locally derived keccak256 of the unsigned envelope; sign as a raw digest. */
  signingDigest: string;
}
/** Inspect the proxy's canonical unsigned type-2 deposit. Rejects signed/other transaction types. */
export function parseEvmUnsignedTransaction(unsignedTransaction: string): EvmDepositInspection {
  const bytes = unhex(unsignedTransaction);
  const reader = new Reader(bytes);
  requireValue(reader.num(1) === 2n, "Only unsigned EIP-1559 deposits are supported");
  const fields = rlp(reader);
  reader.done();
  requireValue(Array.isArray(fields) && fields.length === 9, "Expected unsigned EIP-1559 fields");
  requireValue(
    Array.isArray(fields[8]) && fields[8].length === 0,
    "Proxy deposits require an empty access list",
  );
  const chainId = evmInt(fields[0]),
    nonce = evmInt(fields[1]);
  const maxPriorityFeePerGas = evmInt(fields[2]),
    maxFeePerGas = evmInt(fields[3]),
    gasLimit = evmInt(fields[4]);
  requireValue(chainId > 0n && maxPriorityFeePerGas <= maxFeePerGas);
  const messengerAddress = hex(leaf(fields[5], 20)),
    value = evmInt(fields[6]),
    data = leaf(fields[7]);
  const signatures = [
    "deposit(bytes32,address,uint256,bool)",
    "depositWithPermit(bytes32,address,uint256,uint256,uint8,bytes32,bytes32,bool)",
    "depositETH(bytes32,bool)",
  ];
  const methodIndex = signatures.findIndex(
    (s) => hex(keccak_256(new TextEncoder().encode(s)).slice(0, 4)) === hex(data.slice(0, 4)),
  );
  requireValue(methodIndex >= 0, "Unknown Messenger method");
  const count = [4, 8, 2][methodIndex] ?? 0;
  requireValue(data.length === 4 + count * 32, "Invalid Messenger calldata length");
  const arg = (i: number) => data.slice(4 + i * 32, 36 + i * 32);
  const recipient = hex(arg(0)),
    flag = integer(arg(count - 1));
  requireValue(flag <= 1n, "Invalid recipientIsContract");
  let tokenAddress: string | null = null;
  let amount = value;
  let permit: EvmDepositInspection["permit"];
  if (methodIndex !== 2) {
    requireValue(
      value === 0n &&
        arg(1)
          .slice(0, 12)
          .every((b) => b === 0),
    );
    tokenAddress = hex(arg(1).slice(12));
    amount = integer(arg(2));
    if (methodIndex === 1) {
      requireValue(integer(arg(4)) <= 255n, "Invalid permit v");
      permit = {
        deadline: integer(arg(3)),
        v: Number(integer(arg(4))),
        r: hex(arg(5)),
        s: hex(arg(6)),
      };
    }
  }
  return {
    type: 2,
    chainId,
    nonce,
    messengerAddress,
    value,
    gasLimit,
    maxFeePerGas,
    maxPriorityFeePerGas,
    estimatedNetworkFee: gasLimit * maxFeePerGas,
    data: hex(data),
    method: (["deposit", "depositWithPermit", "depositETH"] as const)[methodIndex] ?? "deposit",
    recipient,
    recipientIsContract: flag === 1n,
    tokenAddress,
    amount,
    ...(permit ? { permit } : {}),
    signingDigest: hex(keccak_256(bytes)),
  };
}

export interface FuelInputInspection {
  type: "coin" | "contract" | "message";
  owner?: string;
  amount?: bigint;
  assetId?: string;
  contractId?: string;
  witnessIndex?: number;
}
export interface FuelOutputInspection {
  type: "contract" | "change" | "variable";
  /** Variable recipients/assets and Change/Variable amounts are execution results, not signed guarantees. */
  to?: string;
  amount?: bigint;
  assetId?: string;
  inputIndex?: number;
}
export interface FuelWithdrawalInspection {
  assetId: string;
  /** Asset sub-ID encoded in the Asset Registry call. */
  assetSubId: string;
  assetRegistryContractId: string;
  destinationChainId: number;
  recipient: string;
  grossAmount: bigint;
  /** Fee quote embedded in the call, in Fuel asset base units. Execution may charge a different fee within contract tolerance. */
  bridgeFee: bigint;
  /** grossAmount - bridgeFee: expected, not guaranteed, delivery amount. */
  netAmount: bigint;
  networkFee: { maxFee: bigint };
  expirationBlockHeight: number;
  scriptGasLimit: bigint;
  policies: Record<string, bigint>;
  inputs: FuelInputInspection[];
  outputs: FuelOutputInspection[];
  /** Locally computed using the caller-supplied, independently trusted Fuel chain ID. */
  transactionId: string;
}

/** Inspect the pinned proxy single-call script, not arbitrary Fuel programs.
 * Parsing is not a safety approval: compare recipients, amounts, contracts, inputs and outputs
 * with independently trusted expectations before signing transactionId. No RPC calls are made.
 * @param fuelChainId Independently trusted chain ID used to derive the signing ID.
 * @param fuelMaxInputs Independently trusted consensus txParameters.maxInputs, needed to locate script data in VM memory.
 */
export function parseFuelUnsignedTransaction(
  unsignedTransaction: string,
  fuelChainId: bigint,
  fuelMaxInputs: number,
): FuelWithdrawalInspection {
  requireValue(
    Number.isSafeInteger(fuelMaxInputs) && fuelMaxInputs > 0 && fuelMaxInputs <= 65535,
    "Invalid Fuel consensus maxInputs",
  );
  const bytes = unhex(unsignedTransaction),
    normalized = bytes.slice(),
    r = new Reader(bytes);
  const zero = (start: number, size: number) => normalized.fill(0, start, start + size);
  requireValue(r.num() === 0n, "Expected Fuel Script transaction");
  const scriptGasLimit = r.num();
  r.take(32);
  zero(16, 32);
  const scriptLength = r.count(),
    dataLength = r.count(),
    mask = r.count(63);
  const inputCount = r.count(),
    outputCount = r.count(),
    witnesses = r.count();
  requireValue(
    inputCount > 0 && inputCount <= fuelMaxInputs && witnesses === 1,
    "Expected supported inputs and one unsigned owner witness",
  );
  const script = r.padded(scriptLength),
    data = r.padded(dataLength);
  const policies: Record<string, bigint> = {};
  ["tip", "witnessLimit", "maturity", "maxFee", "expiration", "owner"].forEach((name, i) => {
    if (mask & (1 << i)) policies[name] = r.num();
  });
  requireValue(
    policies.maxFee !== undefined &&
      policies.expiration !== undefined &&
      policies.expiration <= 0xffffffffn,
  );
  const inputs: FuelInputInspection[] = [];
  for (let i = 0; i < inputCount; i++) {
    const kind = r.count(2),
      start = r.offset;
    if (kind === 1) {
      r.take(32);
      r.count(65535);
      r.take(64);
      r.count(0xffffffff);
      r.count(65535);
      zero(start, 120);
      inputs.push({ type: "contract", contractId: hex(r.take(32)) });
    } else {
      let owner: string, amount: bigint, assetId: string | undefined;
      if (kind === 0) {
        r.take(32);
        r.count(65535);
        owner = hex(r.take(32));
        amount = r.num();
        assetId = hex(r.take(32));
        const pointer = r.offset;
        r.count(0xffffffff);
        r.count(65535);
        zero(pointer, 16);
      } else {
        r.take(32);
        owner = hex(r.take(32));
        amount = r.num();
        r.take(32);
      }
      const witnessIndex = r.count(65535);
      requireValue(witnessIndex === 0);
      const gasOffset = r.offset;
      requireValue(r.num() === 0n, "Predicates are unsupported");
      zero(gasOffset, 8);
      if (kind === 2) requireValue(r.num() === 0n, "Message data is unsupported");
      requireValue(r.num() === 0n && r.num() === 0n, "Predicates are unsupported");
      inputs.push({
        type: kind === 0 ? "coin" : "message",
        owner,
        amount,
        ...(assetId ? { assetId } : {}),
        witnessIndex,
      });
    }
  }
  const outputs: FuelOutputInspection[] = [];
  for (let i = 0; i < outputCount; i++) {
    const kind = r.count(3);
    requireValue(kind !== 0, "Coin outputs are unsupported in proxy withdrawals");
    if (kind === 1) {
      const inputIndex = r.count();
      requireValue(inputs[inputIndex]?.type === "contract");
      zero(r.offset, 64);
      r.take(64);
      outputs.push({ type: "contract", inputIndex });
    } else {
      const start = r.offset,
        to = hex(r.take(32)),
        amount = r.num(),
        assetId = hex(r.take(32));
      if (kind === 2) zero(start + 32, 8);
      if (kind === 3) zero(start, 72);
      outputs.push({
        type: kind === 2 ? "change" : "variable",
        to,
        amount,
        assetId,
      });
    }
  }
  const witnessOffset = r.offset;
  requireValue(r.num() === 0n, "Expected empty unsigned witness");
  r.done();
  zero(88, 8);
  const d = new Reader(data);
  const grossAmount = d.num(),
    assetId = hex(d.take(32)),
    assetRegistryContractId = hex(d.take(32));
  const selectorPointer = d.count(0x3ffff + 88),
    argsPointer = d.count(0x3ffff + 256);
  const selector = new TextEncoder().encode("withdraw_via_fast_bridge_with_fee");
  requireValue(
    d.num() === BigInt(selector.length) && hex(d.take(selector.length)) === hex(selector),
    "Unknown Fuel call",
  );
  requireValue(
    argsPointer === selectorPointer + 8 + selector.length,
    "Invalid Fuel argument pointer",
  );
  const amountOffset = selectorPointer - 88;
  // VM memory prefix = 72 + maxInputs * 40; Script header = 96; CALL script = 24.
  requireValue(
    amountOffset === 192 + 40 * fuelMaxInputs,
    "Fuel call pointer does not address this transaction's script data",
  );
  requireValue(amountOffset >= 0 && amountOffset + 40 <= 0x3ffff);
  const expected = [
    0x72400000 | (amountOffset + 40),
    0x72440000 | amountOffset,
    0x5d451000,
    0x72480000 | (amountOffset + 8),
    0x2d41148a,
    0x24040000,
  ];
  const s = new Reader(script);
  requireValue(
    script.length === 24 && expected.every((v) => s.num(4) === BigInt(v)),
    "Unsupported Fuel script",
  );
  const assetSubId = hex(d.take(32)),
    destinationChainId = Number(d.num(4)),
    paddedRecipient = d.take(32),
    bridgeFee = d.num();
  d.done();
  requireValue(paddedRecipient.slice(0, 12).every((b) => b === 0) && grossAmount >= bridgeFee);
  requireValue(
    paddedRecipient.slice(12).some((b) => b !== 0),
    "Zero withdrawal recipient",
  );
  requireValue(
    inputs.some((i) => i.contractId === assetRegistryContractId),
    "Missing Asset Registry input",
  );
  const signingBytes = new Uint8Array(8 + witnessOffset);
  signingBytes.set(word(fuelChainId));
  signingBytes.set(normalized.slice(0, witnessOffset), 8);
  return {
    assetId,
    assetSubId,
    assetRegistryContractId,
    destinationChainId,
    recipient: hex(paddedRecipient.slice(12)),
    grossAmount,
    bridgeFee,
    netAmount: grossAmount - bridgeFee,
    networkFee: { maxFee: policies.maxFee },
    expirationBlockHeight: Number(policies.expiration),
    scriptGasLimit,
    policies,
    inputs,
    outputs,
    transactionId: hex(sha256(signingBytes)),
  };
}

File: sdks/python/src/o2_sdk/bridge/inspection.py
"""Offline inspection of the proxy's pinned transaction formats, not a safety approval.

No RPC calls, HMAC verification, or personal-sign hashing. Compare parsed intent
with independently trusted expectations before signing the locally derived digest.
"""

from __future__ import annotations

import base64
import hashlib
import json
import re
from dataclasses import dataclass

from Crypto.Hash import keccak


def _check(ok: object, message: str = "Invalid bridge transaction") -> None:
    if not ok:
        raise ValueError(message)


def _unhex(value: str) -> bytes:
    _check(len(value) <= 32770 and re.fullmatch(r"0x(?:[\da-fA-F]{2})+", value))
    return bytes.fromhex(value[2:])


def _hex(value: bytes) -> str:
    return "0x" + value.hex()


def _keccak(value: bytes) -> str:
    return _hex(keccak.new(digest_bits=256, data=value).digest())


@dataclass(frozen=True)
class PreparationProofClaims:
    """UNAUTHENTICATED claims. expires_at is Unix seconds; expired proofs still parse."""

    version: int
    key_id: str
    expires_at: int
    signer: str


def parse_preparation_proof(proof: str) -> PreparationProofClaims:
    """Decode claims only. Does not authenticate the signer or bind a transaction."""
    _check(len(proof) <= 2048, "Preparation proof too large")
    parts = proof.split(".")
    _check(len(parts) == 2, "Invalid preparation proof")

    def decode(value: str) -> bytes:
        _check(re.fullmatch(r"[A-Za-z0-9_-]+", value), "Invalid preparation proof encoding")
        data = base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))
        _check(
            base64.urlsafe_b64encode(data).decode().rstrip("=") == value,
            "Noncanonical preparation proof encoding",
        )
        return data

    _check(len(decode(parts[1])) == 32, "Invalid preparation proof MAC length")
    c = json.loads(decode(parts[0]).decode("utf-8"))
    _check(
        isinstance(c, dict) and set(c) == {"version", "keyId", "expiresAt", "signer"},
        "Invalid preparation proof claims",
    )
    _check(
        type(c["version"]) is int
        and c["version"] == 1
        and isinstance(c["keyId"], str)
        and c["keyId"]
        and type(c["expiresAt"]) is int
        and 0 < c["expiresAt"] <= 9007199254740991
        and isinstance(c["signer"], str)
        and re.fullmatch(r"0x([0-9a-f]{40}|[0-9a-f]{64})", c["signer"]),
        "Invalid preparation proof claims",
    )
    return PreparationProofClaims(c["version"], c["keyId"], c["expiresAt"], c["signer"])


class _Reader:
    def __init__(self, data: bytes):
        self.data = data
        self.offset = 0

    def take(self, size: int) -> bytes:
        _check(0 <= size <= len(self.data) - self.offset, "Truncated transaction")
        result = self.data[self.offset : self.offset + size]
        self.offset += size
        return result

    def num(self, size: int = 8) -> int:
        return int.from_bytes(self.take(size), "big")

    def count(self, maximum: int = 16384) -> int:
        result = self.num()
        _check(result <= maximum, "Unsupported transaction size or field")
        return result

    def padded(self, size: int) -> bytes:
        result = self.take(size)
        _check(not any(self.take(-size % 8)), "Nonzero transaction padding")
        return result

    def done(self) -> None:
        _check(self.offset == len(self.data), "Trailing transaction data")


def _rlp(reader: _Reader, depth: int = 0) -> bytes | list:
    _check(depth < 4, "RLP nesting too deep")
    tag = reader.num(1)
    if tag < 128:
        return bytes([tag])
    is_list = tag >= 192
    short, long = (192, 247) if is_list else (128, 183)
    length = tag - short
    if tag > long:
        encoded = reader.take(tag - long)
        length = int.from_bytes(encoded, "big")
        _check(encoded[0] != 0 and 56 <= length <= 16384)
    data = reader.take(length)
    if not is_list:
        _check(length != 1 or data[0] >= 128, "Noncanonical RLP")
        return data
    children = _Reader(data)
    result = []
    while children.offset < len(data):
        result.append(_rlp(children, depth + 1))
    return result


@dataclass(frozen=True)
class EvmPermitInspection:
    deadline: int
    v: int
    r: str
    s: str


@dataclass(frozen=True)
class EvmDepositInspection:
    type: int
    chain_id: int
    nonce: int
    messenger_address: str
    value: int
    gas_limit: int
    max_fee_per_gas: int
    max_priority_fee_per_gas: int
    estimated_network_fee: int
    """gas_limit * max_fee_per_gas, in wei: cap, not actual fee. Excludes rollup L1 data fees."""
    data: str
    method: str
    recipient: str
    recipient_is_contract: bool
    token_address: str | None
    amount: int
    """EVM token base units or wei, not API/Fuel base units."""
    permit: EvmPermitInspection | None
    signing_digest: str


def parse_evm_unsigned_transaction(unsigned_transaction: str) -> EvmDepositInspection:
    """Parse only canonical unsigned EIP-1559 proxy deposits and derive their raw signing digest."""
    raw = _unhex(unsigned_transaction)
    reader = _Reader(raw)
    _check(reader.num(1) == 2, "Only unsigned EIP-1559 deposits are supported")
    fields = _rlp(reader)
    reader.done()
    _check(isinstance(fields, list) and len(fields) == 9, "Expected unsigned EIP-1559 fields")
    assert isinstance(fields, list)
    _check(fields[8] == [], "Proxy deposits require an empty access list")

    def leaf(index: int, size: int | None = None) -> bytes:
        value = fields[index]
        _check(
            isinstance(value, bytes) and (size is None or len(value) == size), "Invalid EVM field"
        )
        assert isinstance(value, bytes)
        return value

    def number(index: int) -> int:
        data = leaf(index)
        _check(len(data) <= 32 and (not data or data[0] != 0), "Noncanonical EVM integer")
        return int.from_bytes(data, "big")

    chain_id, nonce, priority, fee, gas = (number(i) for i in range(5))
    _check(chain_id > 0 and priority <= fee)
    messenger, value, data = _hex(leaf(5, 20)), number(6), leaf(7)
    signatures = [
        "deposit(bytes32,address,uint256,bool)",
        "depositWithPermit(bytes32,address,uint256,uint256,uint8,bytes32,bytes32,bool)",
        "depositETH(bytes32,bool)",
    ]
    indexes = [
        i for i, sig in enumerate(signatures) if _keccak(sig.encode())[2:10] == data[:4].hex()
    ]
    _check(indexes, "Unknown Messenger method")
    index = indexes[0]
    count = [4, 8, 2][index]
    _check(len(data) == 4 + count * 32, "Invalid Messenger calldata length")

    def arg(i: int) -> bytes:
        return data[4 + i * 32 : 36 + i * 32]

    def arg_int(i: int) -> int:
        return int.from_bytes(arg(i), "big")

    flag = arg_int(count - 1)
    _check(flag <= 1, "Invalid recipientIsContract")
    token = None
    amount = value
    permit = None
    if index != 2:
        _check(value == 0 and not any(arg(1)[:12]))
        token, amount = _hex(arg(1)[12:]), arg_int(2)
        if index == 1:
            _check(arg_int(4) <= 255, "Invalid permit v")
            permit = EvmPermitInspection(arg_int(3), arg_int(4), _hex(arg(5)), _hex(arg(6)))
    return EvmDepositInspection(
        2,
        chain_id,
        nonce,
        messenger,
        value,
        gas,
        fee,
        priority,
        gas * fee,
        _hex(data),
        ["deposit", "depositWithPermit", "depositETH"][index],
        _hex(arg(0)),
        flag == 1,
        token,
        amount,
        permit,
        _keccak(raw),
    )


@dataclass(frozen=True)
class FuelInputInspection:
    type: str
    owner: str | None = None
    amount: int | None = None
    asset_id: str | None = None
    contract_id: str | None = None
    witness_index: int | None = None


@dataclass(frozen=True)
class FuelOutputInspection:
    """Change/Variable amounts and Variable to/asset_id are unsigned execution results."""

    type: str
    to: str | None = None
    amount: int | None = None
    asset_id: str | None = None
    input_index: int | None = None


@dataclass(frozen=True)
class FuelNetworkFee:
    max_fee: int
    """Cap in Fuel base-asset units, not an actual network fee."""


@dataclass(frozen=True)
class FuelWithdrawalInspection:
    asset_id: str
    asset_sub_id: str
    asset_registry_contract_id: str
    destination_chain_id: int
    recipient: str
    gross_amount: int
    bridge_fee: int
    """Embedded quote in Fuel asset units; execution fee can drift within contract tolerance."""
    net_amount: int
    """Expected gross_amount - bridge_fee, not guaranteed delivery."""
    network_fee: FuelNetworkFee
    expiration_block_height: int
    script_gas_limit: int
    policies: dict[str, int]
    inputs: list[FuelInputInspection]
    outputs: list[FuelOutputInspection]
    transaction_id: str


def parse_fuel_unsigned_transaction(
    unsigned_transaction: str, fuel_chain_id: int, fuel_max_inputs: int
) -> FuelWithdrawalInspection:
    """Inspect the proxy single-call script and derive its ID using an independently trusted chain ID.

    Rejects other scripts, predicates, message data and non-empty witnesses. Inspect
    the returned inputs/outputs too; parsing alone does not mean safe to sign.
    fuel_max_inputs is independently trusted consensus txParameters.maxInputs:
    the VM memory offset depends on it, but it is not encoded in the transaction.
    """
    _check(
        type(fuel_chain_id) is int and 0 <= fuel_chain_id < 1 << 64, "Fuel chain ID must fit u64"
    )
    _check(
        type(fuel_max_inputs) is int and 0 < fuel_max_inputs <= 65535,
        "Invalid Fuel consensus maxInputs",
    )
    raw = _unhex(unsigned_transaction)
    normalized = bytearray(raw)
    r = _Reader(raw)

    def zero(start: int, size: int) -> None:
        normalized[start : start + size] = bytes(size)

    _check(r.num() == 0, "Expected Fuel Script transaction")
    gas = r.num()
    r.take(32)
    zero(16, 32)
    script_len, data_len, mask = r.count(), r.count(), r.count(63)
    input_count, output_count, witnesses = r.count(), r.count(), r.count()
    _check(
        0 < input_count <= fuel_max_inputs and witnesses == 1,
        "Expected supported inputs and one unsigned owner witness",
    )
    script, data = r.padded(script_len), r.padded(data_len)
    policies = {
        name: r.num()
        for i, name in enumerate(
            ["tip", "witnessLimit", "maturity", "maxFee", "expiration", "owner"]
        )
        if mask & (1 << i)
    }
    _check(
        "maxFee" in policies and "expiration" in policies and policies["expiration"] <= 0xFFFFFFFF
    )
    inputs = []
    for _ in range(input_count):
        kind, start = r.count(2), r.offset
        if kind == 1:
            r.take(32)
            r.count(65535)
            r.take(64)
            r.count(0xFFFFFFFF)
            r.count(65535)
            zero(start, 120)
            inputs.append(FuelInputInspection("contract", contract_id=_hex(r.take(32))))
        else:
            asset = None
            if kind == 0:
                r.take(32)
                r.count(65535)
                owner, amount, asset = _hex(r.take(32)), r.num(), _hex(r.take(32))
                pointer = r.offset
                r.count(0xFFFFFFFF)
                r.count(65535)
                zero(pointer, 16)
            else:
                r.take(32)
                owner, amount = _hex(r.take(32)), r.num()
                r.take(32)
            witness = r.count(65535)
            _check(witness == 0)
            gas_offset = r.offset
            _check(r.num() == 0, "Predicates are unsupported")
            zero(gas_offset, 8)
            if kind == 2:
                _check(r.num() == 0, "Message data is unsupported")
            _check(r.num() == 0 and r.num() == 0, "Predicates are unsupported")
            inputs.append(
                FuelInputInspection(
                    "coin" if kind == 0 else "message", owner, amount, asset, witness_index=witness
                )
            )
    outputs = []
    for _ in range(output_count):
        kind = r.count(3)
        _check(kind != 0, "Coin outputs are unsupported in proxy withdrawals")
        if kind == 1:
            index = r.count()
            _check(index < len(inputs) and inputs[index].type == "contract")
            zero(r.offset, 64)
            r.take(64)
            outputs.append(FuelOutputInspection("contract", input_index=index))
        else:
            start = r.offset
            to, amount, asset = _hex(r.take(32)), r.num(), _hex(r.take(32))
            if kind == 2:
                zero(start + 32, 8)
            if kind == 3:
                zero(start, 72)
            outputs.append(
                FuelOutputInspection("change" if kind == 2 else "variable", to, amount, asset)
            )
    witness_offset = r.offset
    _check(r.num() == 0, "Expected empty unsigned witness")
    r.done()
    zero(88, 8)
    d = _Reader(data)
    gross, asset_id, registry = d.num(), _hex(d.take(32)), _hex(d.take(32))
    selector_pointer, args_pointer = d.count(0x3FFFF + 88), d.count(0x3FFFF + 256)
    selector = b"withdraw_via_fast_bridge_with_fee"
    _check(d.num() == len(selector) and d.take(len(selector)) == selector, "Unknown Fuel call")
    _check(args_pointer == selector_pointer + 8 + len(selector), "Invalid Fuel argument pointer")
    offset = selector_pointer - 88
    # VM prefix 72 + maxInputs * 40, Script header 96, CALL script 24.
    _check(
        offset == 192 + 40 * fuel_max_inputs,
        "Fuel call pointer does not address this transaction's script data",
    )
    _check(offset >= 0 and offset + 40 <= 0x3FFFF)
    expected = [
        0x72400000 | (offset + 40),
        0x72440000 | offset,
        0x5D451000,
        0x72480000 | (offset + 8),
        0x2D41148A,
        0x24040000,
    ]
    s = _Reader(script)
    _check(len(script) == 24 and all(s.num(4) == v for v in expected), "Unsupported Fuel script")
    sub_id, chain, recipient, fee = _hex(d.take(32)), d.num(4), d.take(32), d.num()
    d.done()
    _check(not any(recipient[:12]) and gross >= fee)
    _check(any(recipient[12:]), "Zero withdrawal recipient")
    _check(any(i.contract_id == registry for i in inputs), "Missing Asset Registry input")
    tx_id = _hex(
        hashlib.sha256(fuel_chain_id.to_bytes(8, "big") + normalized[:witness_offset]).digest()
    )
    return FuelWithdrawalInspection(
        asset_id,
        sub_id,
        registry,
        chain,
        _hex(recipient[12:]),
        gross,
        fee,
        gross - fee,
        FuelNetworkFee(policies["maxFee"]),
        policies["expiration"],
        gas,
        policies,
        inputs,
        outputs,
        tx_id,
    )
